src/Test/QuickCheck/Classes/ShowRead.hs

```diff
@@ -1,43 +1,88 @@
-{-# LANGUAGE CPP #-}
 {-# LANGUAGE ScopedTypeVariables #-}
 
 {-# OPTIONS_GHC -Wall #-}
 
+{-| Module      : Test.QuickCheck.Classes.ShowRead
+    Description : Properties for testing the interaction between the Show and Read
+                  type classes.
+-}
 module Test.QuickCheck.Classes.ShowRead
   ( showReadLaws
   ) where
 
 import Data.Proxy (Proxy)
-import Test.QuickCheck hiding ((.&.))
-import Test.QuickCheck.Property (Property)
+import Test.QuickCheck
+import Text.Read (readListDefault)
+import Text.Show (showListWith)
 
-#if MIN_VERSION_base(4,6,0)
-import Text.Read (readMaybe)
-#endif
-
-import Test.QuickCheck.Classes.Common (Laws(..))
+import Test.QuickCheck.Classes.Common (Laws(..), ShowReadPrecedence(..),
+  SmallList(..), myForAllShrink)
+import Test.QuickCheck.Classes.Compat (readMaybe)
 
 -- | Tests the following properties:
 --
--- [/Partial Isomorphism/]
---   @'readMaybe' ('show' a) == 'Just' a@
---  
--- /Note:/ When using @base-4.5@ or older, this
--- instead test the following:
+-- [/Partial Isomorphism: 'show' \/ 'read'/]
+--   @'readMaybe' ('show' a) ≡ 'Just' a@
+-- [/Partial Isomorphism: 'show' \/ 'read' with initial space/]
+--   @'readMaybe' (" " ++ 'show' a) ≡ 'Just' a@
+-- [/Partial Isomorphism: 'show' \/ 'read' in context/]
+--   @'readMaybe' (" " ++ 'show' __E__[a]) ≡ 'Just' __E__[a]@
```

Comment:
What is `__E__`?

Reply:
Oh actually that refers to a property that I decided not merging (yet) and therefore also did not complete its documentation. I'll remove it. `__E__` is supposed to be an expression context.